cpp/modules/deck.gl/core/src/arrow/row.h

```diff
@@ -79,18 +110,40 @@ class Row {
   auto _getRowChunkData(const std::shared_ptr<arrow::Table>& table, int64_t rowIndex) const -> std::tuple<int, int64_t>;
 
   auto _getDouble(const std::shared_ptr<arrow::Array>& chunk) const -> std::optional<double>;
-  auto _vector2FromFloatArray(const std::shared_ptr<arrow::FloatArray>& values, int32_t offset, int32_t listSize,
-                              const mathgl::Vector2<float>& defaultValue) const -> mathgl::Vector2<float>;
-  auto _vector2FromDoubleArray(const std::shared_ptr<arrow::DoubleArray>& values, int32_t offset, int32_t listSize,
-                               const mathgl::Vector2<double>& defaultValue) const -> mathgl::Vector2<double>;
-  auto _vector3FromFloatArray(const std::shared_ptr<arrow::FloatArray>& values, int32_t offset, int32_t listSize,
-                              const mathgl::Vector3<float>& defaultValue) const -> mathgl::Vector3<float>;
-  auto _vector3FromDoubleArray(const std::shared_ptr<arrow::DoubleArray>& values, int32_t offset, int32_t listSize,
-                               const mathgl::Vector3<double>& defaultValue) const -> mathgl::Vector3<double>;
-  auto _vector4FromFloatArray(const std::shared_ptr<arrow::FloatArray>& values, int32_t offset, int32_t listSize,
-                              const mathgl::Vector4<float>& defaultValue) const -> mathgl::Vector4<float>;
-  auto _vector4FromDoubleArray(const std::shared_ptr<arrow::DoubleArray>& values, int32_t offset, int32_t listSize,
-                               const mathgl::Vector4<double>& defaultValue) const -> mathgl::Vector4<double>;
+  auto _getListArrayMetadata(const std::shared_ptr<arrow::Array>& array, int64_t index) const
+      -> std::shared_ptr<ListArrayMetadata>;
+
+  template <typename T>
+  auto _getVectorData(const std::string& columnName) const -> std::shared_ptr<std::vector<T>> {
+    if (!this->isValid(columnName)) {
+      return nullptr;
+    }
+
+    auto metadata = this->_getListArrayMetadata(this->_getChunk(columnName), this->_chunkRowIndex);
+
+    auto data = std::make_shared<std::vector<T>>();
+    if (auto doubleArray = std::dynamic_pointer_cast<arrow::DoubleArray>(metadata->values)) {
```

Comment:
Dynamic casts are somewhat expensive as they have to inspect RTTI data. It is rarely an issue, only if done very frequently inside loops - but unfortunately that is the case with all Row methods.

I would strongly prefer that we switched on the arrow type constant here instead.

We should set up a simple benchmark so we can test the row method performance (in particular we need to cache the string value lookups).